### DISCUSSION TITLE
Filament Stripe Webhook failing with 419 Error (Page Expired) despite CSRF exclusion – Need Help

### DISCUSSION THREAD

**Discussion by ansgarscheffold** [filamentphp/filament#14142] [Help] [answered]
### Package

Other

### Package Version

v3.0.0

### How can we help you?

I’m facing a persistent issue with my filament application after migrating to a new server. The problem specifically occurs when handling Stripe webhooks using a checkout session. Despite multiple configurations and attempts to resolve it, the issue remains unresolved. My application is using a checkout session using a StripeController.php and a custom route. Users can buy an item, directly checkout and then the webhook sends back if the checkout was successfull. But everytime the webhook (checkout.session.completed) is sent to my application, it results in an HTTP 419 error with the message “Page Expired.” I checked server logs (access and error logs) and can confirm 419 errors on webhook requests. The filament log file show no error.

Setup:
- Server: Apache with KeyHelp server panel runnning on Ubuntu on my own VPS, PHP 8.3
- SSL: Let’s Encrypt SSL certificates managed via KeyHelp.
- CDN: Cloudflare is used with full DNS setup and proxy enabled.
- Application: Filament v3 running under a subdomain

What i've tried so far:
- Added the /webhook route to the $except array in VerifyCsrfToken.php to exclude it from CSRF verification
- ENV config file: tried modifying SESSION_SAME_SITE and Configured SESSION_DRIVER, verified Stripe secret keys
- Configured Apache to handle forwarded headers from Cloudflare correctly by setting X-Forwarded-Proto and X-Forwarded
- Created cache rules in Cloudflare specifically excluding the /webhook path to ensure no caching issues
- Configured Laravel’s session and cookie settings to support cross-origin requests (SameSite=None and secure cookies enabled).
- Also tried to set folder permissions for storage and bootstrap/cache to 777 to ensure no permission issues are blocking the webhooks, without success
- Disabled the firewall temporarily via KeyHelp to rule out firewall-related problems
- Checked PHP settings such as memory limits, execution times, and post sizes to ensure they are adequate for handling Stripe requests
- Created and configured a custom TrustProxies middleware file to correctly handle proxy headers, ensuring that the app recognizes requests as HTTPS when behind Cloudflare

I appreciate any suggestions from the community. I’m running out of ideas, and I hope someone can provide some insight or suggest a different approach to resolving this.

app/Http/Controllers/StripeController.php:
```
<?php

namespace App\Http\Controllers;

use App\Models\Ecg;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;
use Stripe\Checkout\Session;
use Stripe\Stripe;
use Stripe\Webhook;

class StripeController extends Controller
{
    public function createCheckoutSession(Request $request)
    {
        // Setze den API-Schlüssel über die Config
        Stripe::setApiKey(config('services.stripe.secret'));

        $amount = $request->query('amount');
        $ecgId = $request->query('ecg_id'); // Get the ECG ID from the query parameters

        $session = Session::create([
            'payment_method_types' => ['card'],
            'line_items' => [[
                'price_data' => [
                    'currency' => 'eur',
                    'product_data' => [
                        'name' => 'EKG-Befundung',
                    ],
                    'unit_amount' => $amount * 100, // amount in cents
                ],
                'quantity' => 1,
            ]],
            'mode' => 'payment',
            'success_url' => route('payment.success', ['ecg_id' => $ecgId]),
            'cancel_url' => route('payment.cancel'),
            'metadata' => [
                'ecg_id' => $ecgId, // Store the ECG ID in metadata
            ],
        ]);

        return redirect()->away($session->url);
    }

    public function success(Request $request)
    {
        $ecgId = $request->query('ecg_id');
        return view('payment.success', ['ecg_id' => $ecgId]);
    }

    public function cancel()
    {
        return view('payment.cancel');
    }

    public function handleWebhook(Request $request)
    {
        Stripe::setApiKey(config('services.stripe.secret'));
        $endpointSecret = config('services.stripe.webhook_secret');
        
        $payload = $request->getContent();
        $sigHeader = $request->header('Stripe-Signature');

        try {
            $event = Webhook::constructEvent($payload, $sigHeader, $endpointSecret);
        } catch (\UnexpectedValueException $e) {
            return response()->json(['error' => 'Invalid payload'], 400);
        } catch (\Stripe\Exception\SignatureVerificationException $e) {
            return response()->json(['error' => 'Invalid signature'], 400);
        }

        switch ($event->type) {
            case 'checkout.session.completed':
                $session = $event->data->object;
                $ecgId = $session->metadata->ecg_id;

                $ecg = Ecg::find($ecgId);
                if ($ecg) {
                    $ecg->isPaid = true; // Update the isPaid property
                    $ecg->save();
                }
                break;
            default:
                Log::warning('Unhandled event type: ' . $event->type);
        }

        return response()->json(['status' => 'success']);
    }
}

```

app/Http/Middleware:
```
<?php

namespace App\Http\Middleware;

use Illuminate\Foundation\Http\Middleware\VerifyCsrfToken as Middleware;

class VerifyCsrfToken extends Middleware
{
    /**
     * The URIs that should be excluded from CSRF verification.
     *
     * @var array
     */
    protected $except = [
        '/webhook', 
        'webhook', 
        'webhook/*',
    ];
}
```

routes/web.php:
```
<?php

use App\Http\Controllers\InvoiceController;
use App\Http\Controllers\ReportController;
use App\Http\Controllers\StripeController;
use Illuminate\Support\Facades\Route;

Route::get('/invoices/{id}/download', [InvoiceController::class, 'downloadPdf'])->name('invoices.download');
Route::get('/reports/{id}/download', [ReportController::class, 'downloadReportPdf'])->name('reports.download');

Route::get('/create-checkout-session', [StripeController::class, 'createCheckoutSession'])->name('create.checkout.session');
Route::get('/payment-success', [StripeController::class, 'success'])->name('payment.success');
Route::get('/payment-cancel', [StripeController::class, 'cancel'])->name('payment.cancel');

// Define the POST route for the webhook
Route::post('/webhook', [StripeController::class, 'handleWebhook'])->name('webhook.handle');
```

.env:
```APP_NAME=xxx
APP_ENV=production
APP_KEY=xxx
APP_DEBUG=true
APP_TIMEZONE=Europe/Berlin
APP_URL=https://app.domain.de

APP_LOCALE=de
APP_FALLBACK_LOCALE=de
APP_FAKER_LOCALE=de_DE

APP_MAINTENANCE_DRIVER=file
# APP_MAINTENANCE_STORE=database

STRIPE_SECRET=xxx
STRIPE_WEBHOOK_SECRET=xxx

BCRYPT_ROUNDS=12

LOG_CHANNEL=stack
LOG_STACK=single
LOG_DEPRECATIONS_CHANNEL=null
LOG_LEVEL=debug

DB_CONNECTION=sqlite
# DB_HOST=127.0.0.1
# DB_PORT=3306
# DB_DATABASE=laravel
# DB_USERNAME=root
# DB_PASSWORD=

SESSION_DRIVER=cookie
SESSION_LIFETIME=120
SESSION_DOMAIN=.domain.de
SESSION_SAME_SITE=None
SESSION_SECURE_COOKIE=true
# SESSION_PATH=/

BROADCAST_CONNECTION=log
FILESYSTEM_DISK=local
QUEUE_CONNECTION=sync

CACHE_STORE=database
CACHE_PREFIX=
```

**Comment by leandrocfe** [comment]
This isn't a Filament issue I guess

**Reply by ansgarscheffold** [reply]
I was hoping that someone has an idea how to troubleshoot the issue..

**Comment by ansgarscheffold** [comment, marked as answer]
I was able to find and fix the error myself. Instead of making the Csrf exception in the VerifyCsrfToken.php file, in the latest Laravel it now has to be inserted in the 'bootstrap/app.php' file.
